### ISSUE TITLE
Allow and Prefer PGP Encrypted Credentials Files

### ISSUE THREAD

**Issue by naftulikay** [naftulikay/aws-env#10]
Since having credentials in plain-text on your filesystem is probably a bad idea, `aws-env` should attempt to load encrypted files in the following order:

 - `~/.aws/credentials.gpg`
 - `~/.aws/credentials.pgp`
 - `~/.aws/credentials.asc`

It should then attempt to decrypt these files using `gpg2` or `gpg` into memory, and then do the export thing.

Failing finding an encrypted file, it should then default back to the normal `~/.aws/credentials`.

**Comment by naftulikay** [comment]
Something like this should find the first extant credentials file

```python
import os

first = lambda a: a[0] if len(a) > 0 else None

credential_file = first(
    filter(lambda f: os.path.isfile(f), 
        map(lambda f: os.path.expanduser(f), ["~/.aws/credentials.gpg", "~/.aws/credentials.pgp", "~/.aws/credentials.asc", "~/.aws/credentials])
    )
)

if not credential_file:
    raise Exception("No credential file found.")

if credential_file.endswith((".gpg", ".pgp", ".asc")):
    config_data = decrypt(credential_file)
else:
    with open(credential_file, "r") as f:
        config_data = f.read()

# continue
```

This will get the first file that exists of the mentioned files.

**Comment by naftulikay** [comment]
For decryption, we need to find a `gpg` binary in the path, so something like this:

```python
import os

first = lambda a: a[0] if len(a) > 0 else None
pgp_executable = first(
    filter(lambda f: os.path.isfile(f), [
        os.path.join(p, f) for f in ['gpg2', 'gpg'] for  p in os.environ.get('PATH').split(os.path.pathsep)
    ])
)
```

This does not check for executability, but if `pgp_executable` at the end of this is `None` we cannot decrypt.

**Comment by naftulikay** [comment]
So first, order of operations should be:

 1. Attempt to find a PGP binary. If not present, the search path should only be `["~/.aws/credentials"]`.
 2. Attempt to find the first credentials file present.
    1. If it is encrypted, attempt to decrypt it using the PGP binary found.
    2. Otherwise just read it into memory.
 3. Resume existing logic.

**Comment by cxhercules** [comment]
Also part of problem is supporting current tools that use this, like python boto, and aws cli.

**Comment by naftulikay** [comment]
@hhercules My use case for this is that I'm willing to charge ahead on this without waiting for support from the CLI because environment variables are _always_ supported. Since the primary use case of this tool is to export a given profile to the environment, supporting GnuPG still allows a huge win for me because my credentials are never persisted in plaintext to disk, so they're encrypted at rest even on unencrypted media due to GnuPG.

**Comment by cxhercules** [comment]
charge on ahead then :), 👏 👏 👏

**Comment by naftulikay** [comment]
Latest via #16 supports a complex strategy for encrypted and unencrypted credentials files.

 TL;DR: `aws-env list` will list your profiles, and `aws-env export` will export a profile intelligently; the highest priority profile wins, and priority is decided by string sort of files in `~/.aws/credentials.d`, prefers encrypted over unencrypted files with the same prefix, and `~/.aws/credentials` _always wins_.
